### ISSUE TITLE
revoscalePY codes not working on sql server 2017

### ISSUE THREAD

**Issue by adityapt** [MicrosoftDocs/feedback#434]
# revoscalepy

Be as descriptive as you can with your idea/bug report/question/suggestion.
I tried to replicate the same code provided in 
https://docs.microsoft.com/en-us/machine-learning-server/python-reference/revoscalepy/rx-data-step
in sql server 2017 using revoscalepy and it is not working. 

execute sp_execute_external_script 
@language = N'Python',
@script = N' 
import revoscalepy as rp 
file= rp.RxXdfData("file.xdf") 

def funct(data, cutoff, newcol): 
    data_1 = data 
    data_1[newcol] = data.apply(lambda row: True if row.value > cutoff else False, axis=1) 
    return data_1 

data = rp.rx_data_step(input_data = file, transform_function = funct, transform_objects = {"cutoff":1, "newcol":"column"}) 

OutputDataSet = data.head() ', 
@input_data_1 = N'';

I am also having issues executing other functions in revoscalepy while RevoscaleR works well, I am not sure why revoscalepy is not working well

## Issue Details
Error in execution.  Check the output for more information.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files\Microsoft SQL Server\MSSQL14.MSSQLSERVER\PYTHON_SERVICES\lib\site-packages\revoscalepy\RxSerializable.py", line 254, in rxTransformFunction
    ret = resolve_dataframe(originalTransformFunc(parse_result_dataframe(dataset), context)) 
TypeError: funct() missing 1 required positional argument: 'newcol'

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files\Microsoft SQL Server\MSSQL14.MSSQLSERVER\PYTHON_SERVICES\lib\site-packages\revoscalepy\RxSerializable.py", line 254, in rxTransformFunction
    ret = resolve_dataframe(originalTransformFunc(parse_result_dataframe(dataset), context)) 
TypeError: funct() missing 1 required positional argument: 'newcol'

Msg 39019, Level 16, State 2, Line 1788
An external script error occurred: 
DataStep error: ERROR: The sample data set for the analysis has no variables.Traceback (most recent call last):
  File "<string>", line 5, in <module>
  File "E:\Databases\MSSQL14.MSSQLSERVER\MSSQL\ExtensibilityData\MSSQLSERVER01\7121D995-3E14-4F41-98F0-42DA109B9918\sqlindb.py", line 49, in transform
    data = rp.rx_data_step(input_data = marketing, transform_function = funct, transform_objects = {"cutoff":1, "newcol":"column"})   
  File "C:\Program Files\Microsoft SQL Server\MSSQL14.MSSQLSERVER\PYTHON_SERVICES\lib\site-packages\revoscalepy\etl\RxDataStep.py", line 320, in rx_data_step
    functionname=function_name, params=call_parameters)
  File "C:\Program Files\Microsoft SQL Server\MSSQL14.MSSQLSERVER\PYTHON_SERVICES\lib\site-packages\revoscalepy\RxSerializable.py", line 291, in rx_native_call
    ret = px_call(functionname, params)
RuntimeError: revoscalepy function failed.

I am not sure what is going wrong here. Can someone please help me out with this?

**Comment by Powerhelmsman** [comment]
@HeidiSteen Can you take a look at this issue?

**Comment by adityapt** [comment]
Hi @HeidiSteen any update on this issue? I am currently translating all the Revoscalepy codes into RevoscaleR as the Revoscalepy codes aren't working well. I also noticed that RevoPemaR codes are not working right on SQL Server while they work well in my local Micrisoft R installation (not the server Microsoft R)

**Comment by HeidiSteen** [comment]
RevoPemaR is not supported on SQL in-database.  For revoscalepy, you mention that the example in rx-data-step doesn't work in SQL.  What error do you get when you run that example (as opposed to your revised example).

**Comment by adityapt** [comment]
This is exactly what I ran:

execute sp_execute_external_script
		@language = N'Python',
		@script = N'
import os
from revoscalepy import RxOptions, RxXdfData, rx_data_step
sample_data_path = RxOptions.get_option("sampleDataDir")
print(sample_data_path)
kyphosis = RxXdfData(os.path.join(sample_data_path, "kyphosis.xdf"))
# Function to label whether the age in month is over 120 months.
month_limit = 120
new_col_name = "Over10Yr"
def transformFunc(data, cutoff, new_col_name):
    ret = data
    ret[new_col_name] = data.apply(lambda row: True if row.Age > cutoff else False, axis=1)
    return ret

kyphosis_df = rx_data_step(input_data=kyphosis, transform_function=transformFunc,
                           transform_objects={"cutoff": month_limit, "new_col_name": new_col_name})
kyphosis_df.head()		
		',
		@input_data_1 = N''

**Comment by HeidiSteen** [comment]
The code example for rx_data_step works fine for me so this doesn't seem to be a content issue.  I did a web search on your error.  It would appear that your error is in how you are calling the variable. Can you see if these links help?  

1- https://stackoverflow.com/questions/39438509/typeerror-fun-missing-1-required-positional-argument-link 

2 - https://stackoverflow.com/questions/47901868/typeerror-restaurantmenu-missing-1-required-positional-argument-restaurant

**Comment by HeidiSteen** [comment]
Since this is not a content issue, if you could please close it to remove it from our queue?  Thank you!

#please-close

**Comment by HeidiSteen** [comment]
I forgot to mention that RevoPemaR is mostly for Hadoop and I don't think it's coming to SQL any time soon. For parallelization, you should check out rx_exec and rx_exec_by.   If your time frame is flexible, there is additional functionality on the roadmap for parallelization, which you can experiment with in an upcoming release, hopefully in the next several months. 

Links for functions:  
1 - https://docs.microsoft.com/en-us/machine-learning-server/python-reference/revoscalepy/rx-exec
2 - https://docs.microsoft.com/en-us/machine-learning-server/python-reference/revoscalepy/rx-exec-by

**Comment by Powerhelmsman** [comment]
@adityapt Please take a look at HeidiSteen's reply. I will close the issue so far based on HeidiSteen's comment. If you have other question related to this issue, we can reopen it and continue the discussion.

**Comment by Powerhelmsman** [comment]
@HeidiSteen Thanks for your help. I closed the issue as you suggested.

**Comment by adityapt** [comment]
Thank you very much @HeidiSteen regarding the rx_exec and rx_exec_by functions. I will try them out and see if they help me in the same I was trying to do using RevoPemaR. As you also mentioned, I can definitely try the next release regarding the parallelization.

However, I am still unable to figure out the issue with rx_data_step in Python. I tried creating a user defined class with the function and passing on the variables. I tried a different function altogether just to create a new column with its entries as 1 and it still doesn't work and in this case, it wasn't even recognizing the parameters that I was passing to the function.

Couple of things i observed:

If i create a function with only one input parameter apart from the data parameter, I get this error:
TypeError: unhashable type: 'dict'

If I create a function with two input parameters apart from the data parameter, i get an error that says the second additional parameter is missing.
Ex:
TypeError: transformFunc() missing 1 required positional argument: 'new_col_name'

I tried these on the Python_services interpreter to see if running on SQL is causing this issue but it still throws me the same errors.
Can there be some issue regarding the installation of package that can create this issue?

**Comment by adityapt** [comment]
Thank you very much for providing the link. After you said that it cannot be package issue as functions can be called, I tried calling all the functions available in RevoscalePy to see if some of them don't work. Surprisingly I saw that I couldn't call lot of them. Looks like it was actually issue with the package version.  I asked our Tech to look into it and they fixed it.

Its because of the known python scripting issues on RTM version of sql server 2017 and they applied CU 7 updates that fixed the issue of RevoscalePY - rx_data_step and other functions and now they are working fine.

Thank you very much for all the help.
